docs/specs/sms_gateway.md

```diff
@@ -0,0 +1,73 @@
+# SMS Gateway
+
+This document describes configuration of sms gateway.
+
+- [Using Custom SMS Gateway](#using-custom-sms-gateway)
+  - [Configuration](#configuration)
+  - [Webhook Signature](#webhook-signature)
+  - [Webhook Body](#webhook-body)
+  - [Deno Script](#deno-script)
+
+## Using Custom SMS Gateway
+
+When using custom sms gateway, `messaging.sms_provider` must be set to value `custom` in configuration. A secret with key `sms.custom` must also exist in secret configuration.
+
+### Configuration
+
+The following section must be included in `authgear.yaml`:
+
+```
+messaging:
+  sms_provider: custom
+```
+
+The following section must be included in `authgear.secret.yaml`:
+
+```yaml
+secrets:
+  - data:
+      url: "https://myapp.com/send_sms"
+      secret: "somesecretkey"
+    key: sms.custom
+```
+
+When using a deno script, secret can be unspecified:
+
+```yaml
+secrets:
+  - data:
+      url: "authgeardeno:///deno/randomstring.ts"
+    key: sms.custom
+```
+
+### Webhook Signature
+
+Each request is signed with the secret key provided in the `sms.custom` section with `secret` key. The developer must validate the signature and reject requests with invalid signature to ensure the request originates from Authgear.
+
+The signature is calculated as the hex encoded value of HMAC-SHA256 of the request body.
+
+The signature is included in the header `x-authgear-body-signature:`.
+
+### Webhook Body
+
+The body is in json format with the following fields:
+
+- `to`: The recipient of the sms.
+- `from`: The sender of the sms.
+- `body`: The body of the sms.
+
+Example:
+
+```json
+{
+  "to": "+85298765432",
+  "from": "+85298765432",
+  "body": "You otp is 123456"
+}
+```
+
+### Deno Script
+
+When `url` in config is an URL with scheme `authgeardeno://`, the referenced deno module will be executed instead of sending a webhook request.
```

Comment:
Just put a link here to reference to the definition of a deno script, instead of repeating the definition here.

Reply:
Maybe I create one more file to explain deno script module and referernce it in both document.